### PR TITLE
Fix flex overflow

### LINKED ISSUE #1618: Can't use IS_MODULE in script (or flex scanner push-back overflow)
### What reproduces the bug?
I've trying to fuzz using libfuzzer and found an interesting case.

```
% sudo ./src/bpftrace -e "BEGIN { IS_MODULE }" 
flex scanner push-back overflow
```

Only `IS_MODULE` has this problem.  After some investigation, apparently, the following part is related.

https://github.com/iovisor/bpftrace/blob/330953c11190c23a12f98d374d1b5eb6d7aad1fe/src/lexer.l#L182-L199

and in this case, `yytext` is `IS_MODULE` and `s` is `(option)__is_defined(option##_MODULE)`.  I'm not sure why the macro is defined nor how to fix this.

### bpftrace --info
```
System
  OS: Linux 5.9.0+ #24 SMP Tue Nov 10 16:05:24 JST 2020
  Arch: x86_64

Build
  version: v0.11.0-240-g3309
  LLVM: 11.0.1
  foreach_sym: yes
  unsafe uprobe: no
  bfd: yes
  bpf_attach_kfunc: yes
  bcc_usdt_addsem: no
  bcc bpf_attach_uprobe refcount: no
  libbpf: yes
  libbpf btf dump: yes
  libbpf btf dump type decl: yes

Kernel helpers
  probe_read: yes
  probe_read_str: yes
  probe_read_user: yes
  probe_read_user_str: yes
  probe_read_kernel: yes
  probe_read_kernel_str: yes
  get_current_cgroup_id: yes
  send_signal: yes
  override_return: yes
  get_boot_ns: yes
  dpath: no

Kernel features
  Instruction limit: 1000000
  Loop support: yes
  btf (depends on Build:libbpf): yes
  map batch (depends on Build:libbpf): yes
  uprobe refcount (depends on Build:bcc bpf_attach_uprobe refcount): no

Map types
  hash: yes
  percpu hash: yes
  array: yes
  percpu array: yes
  stack_trace: yes
  perf_event_array: yes

Probe types
  kprobe: yes
  tracepoint: yes
  perf_event: yes
  kfunc: yes

```

### PR DESCRIPTION
**Pull request by mmisono**
Add workaround for flex scanner push-back overflow.

Closes #1618 

##### Checklist

- [ ] Language changes are updated in `docs/reference_guide.md`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [ ] The new behaviour is covered by tests

### CODE REVIEW

**Comment by mmisono** [comment]
@fbs I encountered the following memleaks. Do you think your PRs fix them? Or is this different issue? Maybe my error handling is wrong.

```
% sudo  ./src/bpftrace -e 'BEGIN {IS_MODULE}'
stdin:1:7-17: ERROR: Too deep recursive or long macro is detected: IS_MODULE, (option)__is_defined(option##_MODULE)
BEGIN {IS_MODULE}
      ~~~~~~~~~~
stdin:1:7-17: ERROR: syntax error, unexpected end of file, expecting }
BEGIN {IS_MODULE}
      ~~~~~~~~~~

=================================================================
==18766==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 24 byte(s) in 1 object(s) allocated from:
    #0 0x5a653d in operator new(unsigned long) (/disk/work/bpftrace2/build/src/bpftrace+0x5a653d)
    #1 0xd16dbc in bpftrace::Parser::parse() /disk/work/bpftrace2/build/src/parser.yy:174:97
    #2 0x922add in bpftrace::Driver::parse() /disk/work/bpftrace2/src/driver.cpp:49:10
    #3 0xa4704f in main /disk/work/bpftrace2/src/main.cpp:733:16
    #4 0x7fffed57e0b2 in __libc_start_main /build/glibc-ZN95T4/glibc-2.31/csu/../csu/libc-start.c:308:16

Direct leak of 24 byte(s) in 1 object(s) allocated from:
    #0 0x5a653d in operator new(unsigned long) (/disk/work/bpftrace2/build/src/bpftrace+0x5a653d)
    #1 0xd1e62c in bpftrace::Parser::parse() /disk/work/bpftrace2/build/src/parser.yy:239:83
    #2 0x922add in bpftrace::Driver::parse() /disk/work/bpftrace2/src/driver.cpp:49:10
    #3 0xa4704f in main /disk/work/bpftrace2/src/main.cpp:733:16
    #4 0x7fffed57e0b2 in __libc_start_main /build/glibc-ZN95T4/glibc-2.31/csu/../csu/libc-start.c:308:16

Indirect leak of 432 byte(s) in 1 object(s) allocated from:
    #0 0x5a653d in operator new(unsigned long) (/disk/work/bpftrace2/build/src/bpftrace+0x5a653d)
    #1 0xd17203 in bpftrace::Parser::parse() /disk/work/bpftrace2/build/src/parser.yy:177:93
    #2 0x922add in bpftrace::Driver::parse() /disk/work/bpftrace2/src/driver.cpp:49:10
    #3 0xa4704f in main /disk/work/bpftrace2/src/main.cpp:733:16
    #4 0x7fffed57e0b2 in __libc_start_main /build/glibc-ZN95T4/glibc-2.31/csu/../csu/libc-start.c:308:16

Indirect leak of 8 byte(s) in 1 object(s) allocated from:
    #0 0x5a653d in operator new(unsigned long) (/disk/work/bpftrace2/build/src/bpftrace+0x5a653d)
    #1 0xd6820a in __gnu_cxx::new_allocator<bpftrace::ast::AttachPoint*>::allocate(unsigned long, void const*) /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/ext/new_allocator.h:114:27
    #2 0xd68170 in std::allocator_traits<std::allocator<bpftrace::ast::AttachPoint*> >::allocate(std::allocator<bpftrace::ast::AttachPoint*>&, unsigned long) /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/alloc_traits.h:444:20
    #3 0xd67dac in std::_Vector_base<bpftrace::ast::AttachPoint*, std::allocator<bpftrace::ast::AttachPoint*> >::_M_allocate(unsigned long) /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_vector.h:343:20
    #4 0xd66f34 in void std::vector<bpftrace::ast::AttachPoint*, std::allocator<bpftrace::ast::AttachPoint*> >::_M_realloc_insert<bpftrace::ast::AttachPoint* const&>(__gnu_cxx::__normal_iterator<bpftrace::ast::AttachPoint**, std::vector<bpftrace::ast::AttachPoint*, std::allocator<bpftrace::ast::AttachPoint*> > >, bpftrace::ast::AttachPoint* const&) /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/vector.tcc:440:33
    #5 0xd561ec in std::vector<bpftrace::ast::AttachPoint*, std::allocator<bpftrace::ast::AttachPoint*> >::push_back(bpftrace::ast::AttachPoint* const&) /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_vector.h:1195:4
    #6 0xd171ef in bpftrace::Parser::parse() /disk/work/bpftrace2/build/src/parser.yy:174:169
    #7 0x922add in bpftrace::Driver::parse() /disk/work/bpftrace2/src/driver.cpp:49:10
    #8 0xa4704f in main /disk/work/bpftrace2/src/main.cpp:733:16
    #9 0x7fffed57e0b2 in __libc_start_main /build/glibc-ZN95T4/glibc-2.31/csu/../csu/libc-start.c:308:16

SUMMARY: AddressSanitizer: 488 byte(s) leaked in 4 allocation(s).
```

**Comment by fbs** [comment]
I don't have any prs related to this open so it must be something else. Maybe the driver.root isn't deleted properly after the error?

**Comment by mmisono** [comment]
hmm.. deleting `root_` after error does not solve this.

**Comment by mmisono** [comment]
It seems #1651 accidentally close this PR.

File: src/lexer.l [hunk 1]
```diff
@@ -198,14 +199,27 @@ struct|union|enum       yy_push_state(STRUCT, yyscanner); buffer.clear(); struct
...
                             }
                             else
                             {
-                              for (z=strlen(s) - 1; z >= 0; z--)
+                              int leng = yyleng;
+                              std::string original_s(s);
+                              std::string original_yytext(yytext);
+                              for (z=strlen(s) - 1; z >= 0; z--){
+                                if (unput_count >= 1000 || leng <= 0) {
+                                  driver.error(loc, std::string("Too deep recursive or long macro is detected: ")
```
Comment (danobi) [inline src/lexer.l:212]: Nit: both issues seem like macro recursion issues. Maybe just say this instead?
```suggestion
                                  driver.error(loc, std::string("Macro recursion limit reached: ")
```

**Review by danobi — COMMENTED** [review]
Small nit but LGTM

**Comment by danobi** [comment]
> @fbs I encountered the following memleaks. Do you think your PRs fix them? Or is this different issue? Maybe my error handling is wrong.

Maybe yyterminate() prevents some kind of memory cleanup?

**Comment by mmisono** [comment]
> > @fbs I encountered the following memleaks. Do you think your PRs fix them? Or is this different issue? Maybe my error handling is wrong.
> 
> Maybe yyterminate() prevents some kind of memory cleanup?

OK, I'll merge this for now and create an issue with the memleaks.

**Pull request merged**